### ISSUE TITLE
Unable to map route because it doesn't exist (yet it clearly exists)

### ISSUE THREAD

**Issue by jonvandermeulen** [bluemixgaragelondon/cf-blue-green-deploy#58]
In v1.3.0 (this works on v1.1.0, but not on v.1.3.0 after I updated).

`Could not map route - Domain my-app-route.my-domain.io not found`

The plugin appears to be unable to map the "live" route to the `-new` app, giving the error above. When I look at the routes for this space, I definitely see it in the list. 

It should also be noted that (because IT bureaucracy) my app names contain underscores and my routes contain dashes (eg: my_app -> my-app.my-domain.io). When `bgd` is creating the route for the `-new` app, it creates the route as `my_app-new.my-domain.io`. That's not a _technically_  valid hostname, but cloud foundry (at least the version used on predix) doesn't seem to have a problem with it.

**Comment by holly-cummins** [comment]
Oh dear, what a shame. Thanks for letting me know. Version 1.3.0 of the plugin does do 'corrected' domain handling, so it's possible something has gone wrong there. We'll see if we can get a test case in for that scenario.

**Comment by TobyRet** [comment]
We are also affected by this problem.

**Comment by bluemixgaragelondon** [comment]
@TobyRet, have you got similar underscores and dashes in your app names and domains?

**Comment by bluemixgaragelondon** [comment]
What's happening under the covers is that the plugin is calling `map-route`, and that's failing. The call to `map-route` is old, but it's possible that the domain is now being read from the manifest in a way it hadn't before.
We've confirmed locally that the following should work around the problem. Before doing the first deploy, run 
    cf create-domain <your-org> my-app-route.my-domain.io

The person who runs `create-domain` needs to be the org manager, so it's possibly not appropriate for the plugin to do that invocation. The goal would be to get behaviour as close as possible to `cf push`. However, we should consider adding a more helpful error message.

**Comment by holly-cummins** [comment]
I've confirmed, as well as reproducing with the bgd plugin, that I can reproduce a similar issue just with `cf push`. If an app hasn't ever been associated with a domain, the first usage of that domain will give the `domain not found` error. So I guess the next thing we need to know in this case is whether a `cf push -f manifest.yml` succeeds even though, for the same manifest, `cf bgd -f manifest.yml` fails.

**Comment by dharries** [comment]
I think this is similar to https://github.com/bluemixgaragelondon/cf-blue-green-deploy/issues/5.

Maybe we need to enable users to override the default smoke test route.  I think a single option `--smoke-test-route` would work.  The route could be handled the same way as [`route` inside a manifest](https://docs.cloudfoundry.org/devguide/deploy-apps/manifest.html#routes), i.e. it supports setting the domain, the hostname and the path.

**Comment by smitt04** [comment]
Any update on this? I don't think `create-domain` is the right way to go. As the domain should be separate from the hostname.  But the routes section is looking at it as domains instead of hostname + domains

**Comment by holly-cummins** [comment]
Could we see an example manifest, to see how the routes, domains, and hostnames are being defined in the manifest for these cases? For example, in the first comment, it looks like the hostname (`my_app-new`) is being handled separately from the domain (`my-app-route.my-domain.io`), but from your comment, @smitt04, it sounds like they're not?

**Comment by smitt04** [comment]
```
applications:
- name: my-app
  routes:
  - route: my-app.example.io
  instances: 2
  memory: 512M
  disk_quota: 1G
  buildpack: python_buildpack
  stack: cflinuxfs2
  command: uwsgi --http :$PORT --wsgi-file run.py --master --processes 4 --threads 2 2>&1
  env:
    ENV: production
```

With this I get `Could not map route: Domain my-app.example.io does not exist`. The domain part is the `example.io`, while the hostname is `my-app`.  It works if I do `cf map-route example.io --hostname my-app` But not with the blue-green plugin.  If I do the `cf create-domain my-app.example.io` first then it works.  But thats not how it should be.

**Comment by holly-cummins** [comment]
Ahh, I see the issue. Thanks for the extra information, that's really helpful.

**Comment by holly-cummins** [comment]
This should now be fixed, in the staging repository. @smitt04 @jonvandermeulen @TobyRet , to pick up the fix
```
cf uninstall-plugin blue-green-deploy
cf add-plugin-repo bgd-staging https://garage-cf-plugins.eu-gb.mybluemix.net/
cf install-plugin blue-green-deploy -r bgd-staging

**Comment by smitt04** [comment]
@holly-cummins it works, thanks.

**Comment by holly-cummins** [comment]
For completeness, the root issue was that if there's a route defined we treat the route as the domain. So we need to handle that more carefully and correctly work out the domain. The fix was a more correct internal implementation that knows routes and domains are different. (The cf cli model we depend on was missing a lot of that support when the code was written, but it got added later after we raised an issue.)
